### PR TITLE
Implement read caching

### LINKED ISSUE #5: Enhance read performance
read() is currently pretty slow, because the calls from the OS get mapped 1:1 to requests to Google Drive, which means we have a lot of overhead to create the HTTP(s) connections for each read() call.

Instead of only reading exactly what the OS is requesting, we should read a good chunk more (maybe double the size? maybe triple?) and cache the surplus of data in memory or on disk until the next read() call. This should result in much fewer API calls and thus much faster file opening. 

It should be avoided to download the whole file though, since on big files, this could make the performance even worse when we try to download, e.g., 10 Gigs of data, before answering the OS'es request for 64K

### PR DESCRIPTION
**Pull request by nschoellhorn**
### Big news today:
This patch implements pretty solid read caching, relates to and for now closes #5 

With this patch, when a file is opened, it gets split into virtual "chunks", each as big as the configured size (2 MiB per default). Those chunks, or rather their metadata, are saved in the `object_chunk` table. On application start, a background worker is being launched which processes all chunks that are marked as incomplete. It downloads the chunks of data and saves them as separate files in the `objects` cache directory.  

Those chunks are downloaded in order of creation, so from front to back. When `read()` wants to read bytes that belong to a chunk that is yet to be downloaded, it marks the related chunk as `waiting`. The worker, in turn, makes sure to prioritize those waiting chunks before continuing to process all other incomplete chunks. 

At the moment, the cache on disk is not limited, even though this patch also introduces a new config variable `capacity` in the `cache` section. I've opened another issue for this feature since this patch is already quite large.

#### Performance benefits
* Uncached reads, mapped one to one to download requests on Google: ~155 KiB/s on a 140 MiB file, Block Size 4k
* Fully cached reads (all chunks complete on disk): ~79 MiB/s on a 140 MiB file, Block Size 4k
* Uncached reads with prioritized chunked downloading: ~3 MiB/s on a 140 MiB file, Block Size 4k

I am pretty happy with how this performance boost plays out and it should be more than sufficient for now. We might implement compression of the cached object files to improve disk space usage while trading in some of the performance, could be a configurable option as well.

### CODE REVIEW

File: migrations/2021-01-23-172932_create_cache_table/down.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+DROP TABLE object_chunk;
+DROP TABLE object_cache_meta;
```
Comment (Nimelrian) [inline migrations/2021-01-23-172932_create_cache_table/down.sql]: Missing new line at EOF

File: src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        // to read bytes, we first need to figure out in which chunks the requested range is saved
+        let chunks =
+            self.chunk_repository
+                .find_chunks_for_range(file_id.as_str(), byte_from, byte_to);
+
+        let mut buffer = BytesMut::with_capacity((byte_to - byte_from + 1) as usize);
+        // zero out the buffer to make sure its fully initialized (length == capacity)
+        buffer.resize(buffer.capacity(), 0);
+
+        let mut buffer_pointer = 0;
```
Comment (Nimelrian) [inline src/cache.rs]: Maybe cursor instead of pointer?

File: migrations/2021-01-23-172932_create_cache_table/down.sql (hunk 1 above)
Comment (nschoellhorn) [reply migrations/2021-01-23-172932_create_cache_table/down.sql]: Good catch, changed :)

File: src/cache.rs (hunk 2 above)
Comment (nschoellhorn) [reply src/cache.rs]: Sounds good, changed.

**Pull request merged**